### PR TITLE
[FLINK-28475] [Connector/kafka] Stopping offset can be 0

### PR DESCRIPTION
**Pull request by leozhangsr**
## What is the purpose of the change
Stopping offset can be 0.When we make a zero stopping offset empty,it would be serialized to NO_STOPPING_OFFSET, which it is not expected.

## Brief change log
change a stopping offset conditon from '>0' to '>=0'

## Verifying this change
This change is a trivial rework / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

**Review by MartijnVisser — CHANGES_REQUESTED** [review]
@leozhangsr Can you add a test for this situation?

**Comment by leozhangsr** [comment]
> @leozhangsr Can you add a test for this situation?

ok, I will add a test later

**Comment by leozhangsr** [comment]
> @leozhangsr Can you add a test for this situation?

I added a test in KafkaPartitionSplitSerializerTest to make sure KafkaPartitionSplit is serialized correctly.

**Comment by leozhangsr** [comment]
@flinkbot run azure

**Review by MartijnVisser — COMMENTED** [review]
Let's wait for a review from @PatrickRen

File: flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/split/KafkaPartitionSplitSerializerTest.java
Comment (PatrickRen) [inline flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/split/KafkaPartitionSplitSerializerTest.java:30]: I think the expected behavior of `KafkaSourceReader` should be: if the stopping offset of a split is set to 0, no message from that split will be consumed. This test case is not validating the expected behavior but only checks the functionality of split serializer.

**Review by PatrickRen — CHANGES_REQUESTED** [review]
Thanks for the patch @leozhangsr ! I think the test case is not quite make sense. I left some comments below.

File: flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/split/KafkaPartitionSplitSerializerTest.java
Comment (leozhangsr) [reply flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/split/KafkaPartitionSplitSerializerTest.java:30]: yes, that's what we want: if the stopping offset of a split is set to 0, no message from that split will be consumed.
I check this code again, and make a explanation for my change and test case.
As we know, split is defined by the driver, then is serialized and send to task manager, then handled by KafkaPartitionSplitReader.The split reader cosumes messages, and stops at the stopping offset if the stopping offset is set.

To achieve this, following key steps have to be validated:
1、split is correctly serialized and send to split reader.
2、split reader parse the split correctly
3、split reader consumes and stop at the stopping offset.

Step 1 is validated by test case I had make.
Step 2 and 3 , can be validated by KafkaPartitionSplitReaderTest.testHandleSplitChangesAndFetch-assignSplitsAndFetchUntilFinish. This test make sure the split reader stop at the stopping offset.This test case set the stopping offset to 10(NUM_RECORDS_PER_PARTITION).Though the stopping offset is not 0, it still can work well if the split is parse correctly. KafkaPartitionSplitReader.parseStoppingOffsets acquires stoppingOffset to be >= 0,LATEST_OFFSET,COMMITTED_OFFSET. The KafkaPartitionSplit.getStoppingOffset reachs the same conditions after my modification.
Step 2 and 3, is also validated by KafkaPartitionSplitReaderTest.testAssignEmptySplit for empty split situation.Generally when the stopping offset is 0, the starting offset might by 0 too, which means it's a empty split,it should consume nothing  and stop.In this test case, the empty split' starting offset is LATEST_OFFSET, stopping offset is LATEST_OFFSET.

I only add a test case for step 1, since it's nerver tested before.I think the existing tests can already covered the situation(if the stopping offset of a split is set to 0, no message from that split will be consumed).Do you agree? Should I add a test case just for stopping offset = 0? Would it be a little redundance?

Comment (PatrickRen) [reply flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/split/KafkaPartitionSplitSerializerTest.java:30]: Thanks for the reply @leozhangsr. I still think it's necessary to have a case testing stopping offset = 0 (step 3). Those cases you mentioned (`KafkaPartitionSplitReaderTest#testHandleSplitChangesAndFetch`, `#assignSplitsAndFetchUntilFinish` and `testAssignEmptySplit`) don't cover the corner case that having stopping offset = 0 in the split, and that's why we didn't catch this bug in CIs.

**Review by mas-chen — COMMENTED** [review]
minor nit, unit test lgtm

File: flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/split/KafkaPartitionSplitSerializerTest.java
Comment (leozhangsr) [reply flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/split/KafkaPartitionSplitSerializerTest.java:30]: Thanks for your suggestion.I have added a test case in KafkaPartitionSplitReaderTest#testAssignEmptySplit to validate the zero stopping offset situation.Please review this.Thanks.

**Comment by MartijnVisser** [comment]
@flinkbot  run azure

**Comment by leozhangsr** [comment]
image
I have checked the ci result, relative change and test case is passed.The ci failed with other test task.I am not sure how to handle this, any one can help? Thanks!

**Comment by MartijnVisser** [comment]
@flinkbot run azure

**Comment by MartijnVisser** [comment]
> I have checked the ci result, relative change and test case is passed.The ci failed with other test task.I am not sure how to handle this, any one can help? Thanks!

Let's retry to get the build passing with a retry; else we can always consider merging it (after the code review process has been completed)

**Comment by leozhangsr** [comment]
@PatrickRen waiting a review from you

**Review by PatrickRen — APPROVED** [review]
Thanks for the patch @leozhangsr ! LGTM. Could you squash all commits into one? I'll merge it then.

**Comment by leozhangsr** [comment]
> Thanks for the patch @leozhangsr ! LGTM. Could you squash all commits into one? I'll merge it then.

I tried, but it seems that commits can't be squashed by me in this branch or mr.Could you use squash merging(squash and merge option ) ? Or I can squash in a new branch and make new mr, you can merge it then.

**Comment by PatrickRen** [comment]
@leozhangsr You can squash them locally and force push this branch to your repository. Never mind I'll merge it.

**Comment by PatrickRen** [comment]
Squashed and merged to master 55fc7fa95aae4304a8694c03adf79d0a609918b1

**Comment by leozhangsr** [comment]
> Squashed and merged to master [55fc7fa](https://github.com/apache/flink/commit/55fc7fa95aae4304a8694c03adf79d0a609918b1)
> 
> @leozhangsr Could you create another two PRs to backport this patch to release-1.14 and release-1.15?

Glad to help!

[release-1.14]：https://github.com/apache/flink/pull/20387
[release-1.15]：https://github.com/apache/flink/pull/20389

**Pull request closed without merging**
